### PR TITLE
[Core] Adding Global Derivatives

### PR DESCRIPTION
**Pull request by tteschemacher**
Dear all,

I am adding here an interface to collect the global derivatives from the geometry base class. The idea is the following:
Index 0: Location
Index 1: Base Vectors
Index 2: Curvature Vectors
...
Those operations are typically operated within the element. However, those are geometrical, thus I would like to introduce an interface here which can compute those.

The interface is inspired by the NURBS book from Piegl and Tiller.

Thanks for the review!

### CODE REVIEW

**Review by loumalouomega — DISMISSED** [review]
OK

File: kratos/geometries/geometry.h [hunk 1]
```diff
@@ -1724,6 +1724,28 @@ class Geometry
         return rResult;
     }
 
+    /**
+    * @brief This method maps from dimension space to working space and computes the
+    *        number of derivatives at the dimension parameter.
+    * @param LocalCoordinates The local coordinates in dimension space
```
Comment (armingeiser) [inline kratos/geometries/geometry.h]: naming of the parameters does not mach implementation

File: kratos/geometries/geometry.h [hunk 2]
```diff
@@ -1724,6 +1724,28 @@ class Geometry
...
+    * @param Derivative Number of computed derivatives
+    * @return std::vector<array_1d<double, 3>> with the coordinates in working space
+    * @see PointLocalCoordinates
+    */
+    virtual void GlobalDerivatives(
+        std::vector<CoordinatesArrayType>& rGlobalCoordinates,
+        const CoordinatesArrayType& rCoordinates,
+        const SizeType DerivativeOrder) const
+    {
+        KRATOS_ERROR << "Calling GlobalDerivatives within geometry.h."
```
Comment (oberbichler) [inline kratos/geometries/geometry.h]: Would it makes sense to call `Geometry::GlobalCoordinates` if `DerivativeOrder == 0`?

File: kratos/geometries/geometry.h [hunk 3]
```diff
@@ -1724,6 +1724,28 @@ class Geometry
...
+    /**
+    * @brief This method maps from dimension space to working space and computes the
+    *        number of derivatives at the dimension parameter.
+    * @param LocalCoordinates The local coordinates in dimension space
+    * @param Derivative Number of computed derivatives
+    * @return std::vector<array_1d<double, 3>> with the coordinates in working space
+    * @see PointLocalCoordinates
+    */
+    virtual void GlobalDerivatives(
+        std::vector<CoordinatesArrayType>& rGlobalCoordinates,
```
Comment (oberbichler) [inline kratos/geometries/geometry.h]: The order in which the derivatives are stored should be explained in the comments.

File: kratos/geometries/geometry.h (hunk 1 above)
Comment (tteschemacher) [reply kratos/geometries/geometry.h]: Thanks, I tried to be more coherent now ;)

File: kratos/geometries/geometry.h (hunk 3 above)
Comment (tteschemacher) [reply kratos/geometries/geometry.h]: Thanks for the comment. I put it accordingly.

File: kratos/geometries/geometry.h (hunk 2 above)
Comment (tteschemacher) [reply kratos/geometries/geometry.h]: I think you're right. I put also DerivativeOrder == 1 as this information can be easily computed.

File: kratos/geometries/geometry.h [hunk 4]
```diff
@@ -1724,6 +1745,134 @@ class Geometry
...
+
+            rGlobalCoordinates[0] = global_coordinates;
+            return;
+        }
+
+        if (DerivativeOrder == 1)
+        {
+            const double local_space_dimension = LocalSpaceDimension();
+
+            if (rGlobalCoordinates.size() != 1)
```
Comment (oberbichler) [inline kratos/geometries/geometry.h]: Is this correct?

File: kratos/geometries/geometry.h (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/geometries/geometry.h]: I think in #4777 we agreed on the interface such function should have...

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
blocking since as far as i can tell this is not in agreement with the decision about the interface we made  in #4777

**Comment by tteschemacher** [comment]
In #4777 we were talking about the shape functions. This one is concerning the evaluated values (location, base vectors, derivatives, ...)
@RiccardoRossi the shape functions are set within #5424

**Comment by RiccardoRossi** [comment]
@tteschemacher to my understanding one can ask for "ShapeFunctionLocalDerivatives" or for "ShapeFunctionDerivatives". The second one are the derivatives in the global system.

the last comment of #4777 was

~~~
@KratosMultiphysics/technical-committee we agreed about having the new interface as follow:

 double D = a.ShapeFunctionDerivatives( derivative_order, node_index, k , l .... );

please note that a is a reference to internal storage of shape functions and derivatives container.
~~~
so as far as i understand this is relevant here

**Comment by tteschemacher** [comment]
Here you can ask for the base vectors at a specific location or integration point. Here, as it is the base class only 0st, and first order is possible. The name might be confusing, however it is adapted from GlobalCoordinates, which is this function for Derivative Order 0.

In #5424 we are adding the container for high order. As soon as this one is set, I will add this interface within a new PR as you mentioned.

**Comment by RiccardoRossi** [comment]
even like this:

* why do you need the container for high order to be different to the one for high order?
* why not calling it ShapeFunctionDerivatives?

**Comment by oberbichler** [comment]
We should pay attention to the distinction between `ShapeFunctionDerivatives` and `GlobalDerivatives`.

We made a small sketch:
GlobalDerivatives-1

In this PR we are talking about `GlobalDerivatives`: The derivatives of an evaluated point (e.g. `c(t)`) of the geometry w.r.t. the local parameters (`t`) and NOT about the derivatives of the shape functions.

**Comment by RiccardoRossi** [comment]
@tteschemacher explained this to me via hangouts, and to be honest i don't think that the name of the function is really explicative. In any case if you have a 3D geometry, what would be the GlobalDerivative(1) ?

**Comment by oberbichler** [comment]
Is a 3D geometry a volume or a geometry in 3D-space?

In both cases (as I unterstand the current implementation) it would return the derivative of the geometry w.r.t. the first local parameter (= tangent vector).

In case you have a curve `c(t) -> d/dt c(t)`
In case you have a surface `s(u, v) -> d/du s(u, v)`
In case you have a volume `v(r, s, t) -> d/dr v(r, s, t)`

In each case it would return a vector with the dimension of the working space.

**Comment by RiccardoRossi** [comment]
this is pretty strange to me. Why _the first_?

Btw if you ask for the Jacobian it will give you something that is very similar to this (first column, the derivative w.r.t r, second derivative w.r.t s , third w.r.t. t

**Comment by tteschemacher** [comment]
For basic geometries the return is similiar to the Jacobian. However, in a different data structure (vector<array_1d>) and not Matrix. For many mathematical functions this is ususally needed.

However, this is first only from the structure. The need of this function comes obvious when you think about embedded geometries (e.g. line on surface). Here, those vectors are also already multiplied with the tangential functions. -> then this function is important

**Comment by RiccardoRossi** [comment]
If you are getting exactly the same as the jacobian, just in a different data structure, then i am not positive about this. 

As i recall (by heart) Jacobian contains for a 2D surface in the 3D space (considering parameters u,v)

  dx/du  dx/dv  
  dy/du  dy/dv
  dy/du  dz/dv

for a 1D matrix in 3D space

  dx/du  
  dy/du
  dy/du

this means that you coud obtain any of the vectors you are interested in by doing,

array<double,3> v,i = col(Jacobian,i)

**Comment by tteschemacher** [comment]
for 1st order yes, but not for embedded geometries. And also some geometries will need different computation. 2nd and higher don't have the possibility to do it like this.

**Comment by oberbichler** [comment]
@RiccardoRossi If you call
```python
global_derivatives = []
GlobalDerivatives(global_derivatives = global_derivatives, local_coordinates = [u, v, 0], derivative_order = 1)
```

then you would get

```
global_derivatives[0] = [ x, y, z ]
global_derivatives[1] = [ dx/du, dy/du, dz/du ]
global_derivatives[2] = [ dx/dv, dy/dv, dz/dv ]
```

Of course `global_derivatives[0]` is equal to `Geometry::GlobalCoordinates` and `global_derivatives[1:3]` is equal to `Geometry::Jacobian`.

**Comment by RiccardoRossi** [comment]
ok, i think i got what you are proposing. so from my point of view:

1 - let's find a better name than "GlobalDerivatives"
2 - I believe that the function should give back a container, to which you can ask for stuff. not directly a vector of vectors (this is what we were telling in #4777). to that container you can ask for the vector you want, by telling the order and the index of the component you derive by.

**Comment by RiccardoRossi** [comment]
what about "GlobalPositionDerivatives"?

aside for the output format,
what happens if i call this function for, say, a triangle?

**Comment by RiccardoRossi** [comment]
The best name i can come up with is "CoordinatesDerivativesInRealSpace"  (no worry, i also do not love it)

aside of the name, which we can eventually (painfully) change at a later time
I need the doc to tell exactly what the return type will be.

so i'd like the docstring to contain something like (i did it as i understood so it may be wrong, please check):

~~~
**1D curve in 3D space** 
[0] -     - [ x, y, z ]
[1] - du  - [ dx/du, dy/du, dz/du ]  - here starts the first derivatives
[2] - duu - [ d^2x/du^2, d^2y/du^2, d^2z/du^2 ]  - here starts the second derivatives

**2D curve in 3D space** 
[0] -      - [ x, y, z ]
[1] - du  - [ dx/du, dy/du, dz/du ]  - here starts the first derivatives
[2] - dv  - [ dx/dv, dy/dv, dz/dv ]
[3] - duu - [ d^2x/du^2, d^2y/du^2, d^2z/du^2 ]  - here starts the second derivatives
[4] - duv - [ d^2x/(dudv), d^2y/(dudv), d^2y/(dudv) ]
[5] - dvv - [ d^2x/dv^2, d^2y/dv^2, d^2z/dv^2 ]

**3D curve in 3D space** 
[0] -        - [ x, y, z ]
[1] - du   - [ dx/du, dy/du, dz/du ]  - here starts the first derivatives
[2] - dv   - [ dx/dv, dy/dv, dz/dv ]
[3] - dt    - [ dx/dt dy/dt, dz/dt ]
[4] - duu - [ d^2x/du^2, d^2y/du^2, d^2z/du^2 ]  - here starts the second derivatives
[5] - duv - [ d^2x/(dudv), d^2y/(dudv), d^2y/(dudv) ]
[6] - dut  - [ d^2x/(dudt), d^2y/(dudt), d^2y/(dudt) ]
[7] - dvv - [ d^2x/dv^2, d^2y/dv^2, d^2z/dv^2 ]
[8] - dvt  - [ d^2x/(dvdt), d^2y/(dvdt), d^2y/(dvdt) ]
[9] - dtt   - ...
~~~

the thing that scares me, is that if i ask for second order, the position number 2 will imply something different in the two cases (and something different again for the 3D case).
**It is to the user to understand what is the correct way to access this... which i see as very error prone**. What i would have loved is a way to automatically recover the, say, derivative w.r.t u,v knowing the local space dimension

**Comment by RiccardoRossi** [comment]
also, what happens on day 1 if we call this function for a, say, triangle? will we implement a function in the base geometry to make it to work or will we need to touch all of the existing geometries to provide this capability?

**Comment by oberbichler** [comment]
my two cents:

> **2D curve in 3D space**

I guess you mean something like `**2-Parameter geometry in 3D space**`.
In case of a volume I am not sure what the common ordering of the derivatives is. For the moment I would keep that information away (same in the other PR).
The rest seems to be correct.

> if i ask for second order, the position number 2 will imply something different in the two cases

Yes. I think you can "solve" this with the container you proposed. But then you will also have a lot of overhead since you need a base class + containers for different geometries.

> What i would have loved is a way to automatically recover the, say, derivative w.r.t u,v knowing the local space dimension

It's just some basic math. I usually do this in a static function.

> what happens on day 1 if we call this function for a, say, triangle?

Looking at the current implementation it will return
- `DerivativeOrder=0`: `Geometry::GlobalCoordinates`
- `DerivativeOrder=1`: `Geometry::GlobalCoordinates` + `Geometry::Jacobian`
- `DerivativeOrder>1`: Error
without touching the existing geometries.

It would make sense to change the last to
- `DerivativeOrder>1`: `Geometry::GlobalCoordinates` + `Geometry::Jacobian` + `ZeroVectors`
in case you only have linear geometries at the moment.

**Comment by tteschemacher** [comment]
GlobalPositionDerivatives

I actually like this name.

DerivativeOrder>1: Error
without touching the existing geometries.

As soon as we have the high order derivatives, this is also done, generically.

**Comment by tteschemacher** [comment]
How about GlobalSpaceDerivatives? That would be a mixture between CoordinatesDerivativesInRealSpace and GlobalPositionDerivatives

**Comment by tteschemacher** [comment]
@RiccardoRossi are you fine with that implementation?

**Comment by tteschemacher** [comment]
Does anyone know how to treat the Appveyor that this compiles in-time?

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee approves the new interface proposed by this PR.

**Comment by tteschemacher** [comment]
ping @RiccardoRossi

**Review by RiccardoRossi — APPROVED** [review]
ok, you still moss the doc for the 3d case, but i understand you will commit it in a different PR

**Comment by tteschemacher** [comment]
thanks. This is already in preparation...

**Pull request merged**